src/app/component/projects/pages/edit/edit.component.ts

```diff
@@ -32,75 +28,55 @@ import { fields } from "./edit.json";
 @Component({
   selector: "app-project-edit",
   template: `
-    <app-wip *ngIf="project">
+    <app-wip *ngIf="models">
       <app-form
         [schema]="schema"
-        [title]="'Edit ' + project.name"
+        [title]="'Edit ' + getProject().name"
         [submitLabel]="'Submit'"
         [submitLoading]="loading"
         (onSubmit)="submit($event)"
       ></app-form>
     </app-wip>
   `
 })
-export class EditComponent extends WithFormCheck(PageComponent)
-  implements OnInit {
-  public loading: boolean;
-  public project: Project;
-  public schema = { model: {}, fields };
-
+export class EditComponent extends EditFormTemplate<Project, FormEvent> {
   constructor(
-    private route: ActivatedRoute,
-    private router: Router,
     private api: ProjectsService,
-    private notification: ToastrService
+    notifications: ToastrService,
+    route: ActivatedRoute,
+    router: Router
   ) {
-    super();
+    super(["project"], fields, notifications, route, router);
   }
 
-  ngOnInit() {
-    const projectModel: ResolvedModel<Project> = this.route.snapshot.data
-      .project;
-
-    if (projectModel.error) {
-      return;
-    }
-
-    this.project = projectModel.model;
-    this.schema.model["name"] = this.project.name;
-    this.schema.model["description"] = this.project.description;
+  public getProject(): Project {
+    return this.models.project as Project;
   }
 
-  /**
-   * Form submission
-   * @param $event Form response
-   */
-  submit($event: any) {
-    const project = new Project({ ...$event, id: this.project.id });
+  preFillForm() {
+    const project = this.getProject();
 
-    this.loading = true;
+    this.schema.model["name"] = project.name;
+    this.schema.model["description"] = project.description;
+  }
 
-    this.api
-      .update(project)
-      .pipe(takeUntil(this.unsubscribe))
-      .subscribe(
-        () => {
-          this.resetForms();
-          this.notification.success("Project was successfully updated.");
-          this.router.navigateByUrl(project.redirectPath());
-        },
-        (err: ApiErrorDetails) => {
-          let errMsg: string;
+  apiUpdate(event: FormEvent) {
+    const project = this.getProject();
+    const updatedProject = new Project({ id: project.id, ...event });
```

Comment:
why are we recreating the model here?

Can't we just push to model that was edited by the form? i.e. `this.models.project`

Reply:
Models are readonly to preserve immutability